### PR TITLE
autodetect, download and install firmware upgrades

### PR DESCRIPTION
**Pull request by proddy**
Further improvements. Now the Download page will check if there is an updated firmware available (for dev or stable depending on which branch you are on). And give the user to download and install automatically.

There is also the option to switch from the Stable branch to the Dev branch (but no way back).

Seems to work on my setup.

### CODE REVIEW

**Comment by proddy** [comment]
And @MichaelDvP I got the same Network Error when uploading using the DragNDrop so will try and fix that too next

**Review by MichaelDvP — CHANGES_REQUESTED** [review]
A lot of changes. I have not checked all.
Appart from the not working log i've found: (not part of this PR)
- the divider in settings, is it intended? https://github.com/proddy/EMS-ESP32/blob/41ec9162fda52b033a4202f459a7c3ad75e506f3/interface/src/app/settings/Settings.tsx#L134
- custom entity adding/editing is still reversed by autoupdate. In this line https://github.com/proddy/EMS-ESP32/blob/41ec9162fda52b033a4202f459a7c3ad75e506f3/interface/src/app/main/CustomEntities.tsx#L57 it should be `if (dialogOpen || numChanges > 0) {`

File: interface/src/app/status/SystemLog.tsx [hunk 1]
```diff
@@ -102,7 +102,6 @@ const SystemLog = () => {
 
   // eslint-disable-next-line @typescript-eslint/unbound-method
   const { onMessage, onError } = useSSE(fetchLogES, {
-    immediate: true,
```
[the remark this answers is not in the document]
Comment (proddy) [reply interface/src/app/status/SystemLog.tsx:105]: good catch. That must be a bug in Alova - I'll report it.

**Comment by proddy** [comment]
> And @MichaelDvP I got the same Network Error when uploading using the DragNDrop so will try and fix that too next

Think I fixed it.

**Comment by proddy** [comment]
> A lot of changes. I have not checked all. Appart from the not working log i've found: (not part of this PR)
> 
> * the divider in settings, is it intended? https://github.com/proddy/EMS-ESP32/blob/41ec9162fda52b033a4202f459a7c3ad75e506f3/interface/src/app/settings/Settings.tsx#L134
> * custom entity adding/editing is still reversed by autoupdate. In this line https://github.com/proddy/EMS-ESP32/blob/41ec9162fda52b033a4202f459a7c3ad75e506f3/interface/src/app/main/CustomEntities.tsx#L57 it should be `if (dialogOpen || numChanges > 0) {`

The divider in Settings was intentional as Upload/Download wasn't really a 'Settings'. But I'll remove it as it does look strange.

**Comment by MichaelDvP** [comment]
Compiled with version set to 3.7.0-dev31. The update-button with link to 3.7.0.dev-32 is shown, but clicking does only a restart.

**Comment by MichaelDvP** [comment]
Hm, status page is workingagain, it was blocked by the update process. 
Log from terminal: repeated every few milliseconds: `Firmware upload failed - HTTP code 4294967295`

**Comment by proddy** [comment]
> damn. both those are bugs - I'll fix

fixed. hopefully.

**Comment by MichaelDvP** [comment]
For me the download did not start, i tested on a esp32-16M without PSRAM, so also no https. 
This stops here https://github.com/proddy/EMS-ESP32/blob/71d90d6416ea257846d992b09b93e7898ab4fd89/src/system.cpp#L1877
but the function just returned false, but does not set `upload_status(false);`, so the upload functions is called every main loop, blocking the status page, there was no reboot at all.
I modufied to 
```
    if (httpCode != HTTP_CODE_OK) {
        LOG_ERROR("Firmware download %s failed - HTTP code %u", saved_url.c_str(), httpCode);
        EMSESP::system_.upload_status(false);
        saved_url.clear(); // prevent from downloading again
        return false;
    }
```
and the log is one line:
`2024-08-30 15:33:56.264 E 16: [system] Firmware download https://github.com/emsesp/EMS-ESP32/releases/download/v3.6.5/EMS-ESP-3_6_5-ESP32-16MB.bin failed - HTTP code 4294967295`
and emsesp keeps responsible.

If i change the url to http:// the http-return is http:301

**Comment by MichaelDvP** [comment]
Update with a ESP32S3-16M+ is working. But webpage is instantly redirected to dashboard while update is working in the background.

**Comment by proddy** [comment]
> Update with a ESP32S3-16M+ is working. But webpage is instantly redirected to dashboard while update is working in the background.

yes, fixed that. Will check-in shortly.

**Comment by MichaelDvP** [comment]
The new dialogs look a bit odd to me.
old / new:
grafik
grafik

**Comment by proddy** [comment]
> The new dialogs look a bit odd to me. old / new: grafik grafik

it's the new Grid component. The box resizes and is responsive. Before those text boxes were too large (e.g. a DeviceID is not 40 chars long). What do you think we should change to make it look nicer?

**Comment by proddy** [comment]
Another 50 files changed, sorry. I think you just need to check the behaviour when EMS-ESP is restarted and what happens when you click on Install

**Comment by MichaelDvP** [comment]
The new restart monitor looks really nice.
Automatic install still does not work on systems without psram (missing https) and failed download from github result in a loop.
I suggest to change emsesp.cpp lines 1676-1678 to
```
    } else if (!system_.uploadFirmwareURL()) { // start an upload from a URL. This is blocking.
        system_.upload_status(false);
    }
```
to stop retrying the GET endless.

The get result is HTTP code 4294967295, or code 301 if i set the url to http://. 
If there is no way to download the bin from github via http, i suggest to remove the install button on systems without psram.

**Comment by proddy** [comment]
> The new restart monitor looks really nice. Automatic install still does not work on systems without psram (missing https) and failed download from github result in a loop. I suggest to change emsesp.cpp lines 1676-1678 to
> 
> ```
>     } else if (!system_.uploadFirmwareURL()) { // start an upload from a URL. This is blocking.
>         system_.upload_status(false);
>     }
> ```
> 
> to stop retrying the GET endless.
> 
> The get result is HTTP code 4294967295, or code 301 if i set the url to http://. If there is no way to download the bin from github via http, i suggest to remove the install button on systems without psram.

Thanks. What I could also do is check if the URL is valid in the Web code and if not, hide the button as you suggested.

I also found some other errors I made in the upload and restart, so will fix those now too.

**Comment by MichaelDvP** [comment]
> What I could also do is check if the URL is valid in the Web code and if not, hide the button as you suggested.

Yes, but the https:// link is valid and the browser can resolve it, but ems-esp-tasmota builds  can not access it. 
For http://  we get 301, checking in terminal with curl it redirects to https.:
```
> GET /emsesp/EMS-ESP32/releases/download/v3.6.5/EMS-ESP-3_6_5-ESP32-16MB.bin HTTP/1.1
< HTTP/1.1 301 Moved Permanently
< Location: https://github.com/emsesp/EMS-ESP32/releases/download/v3.6.5/EMS-ESP-3_6_5-ESP32-16MB.bin
```
Maybe for this cases replace the install button by a download-button.

**Comment by proddy** [comment]
> > What I could also do is check if the URL is valid in the Web code and if not, hide the button as you suggested.
> 
> Yes, but the https:// link is valid and the browser can resolve it, but ems-esp-tasmota builds can not access it. For http:// we get 301, checking in terminal with curl it redirects to https.:
> 
> ```
> > GET /emsesp/EMS-ESP32/releases/download/v3.6.5/EMS-ESP-3_6_5-ESP32-16MB.bin HTTP/1.1
> < HTTP/1.1 301 Moved Permanently
> < Location: https://github.com/emsesp/EMS-ESP32/releases/download/v3.6.5/EMS-ESP-3_6_5-ESP32-16MB.bin
> ```
> 
> Maybe for this cases replace the install button by a download-button.

The HTTPClient call fails and returns -1. I suspect it's because the Tasmota platform has mbed SSL removed which is needed. I'll see if I can find a workaround, otherwise as you said, disabled auto-download for Tasmota builds.

**Comment by mattreim** [comment]
Update German translation: 

```lua
  WAIT_FIRMWARE: 'Die Firmware wird hochgeladen und installiert', 

  RESTARTING_PRE: 'Initialisierung', 
  RESTARTING_POST: 'Neuladen' 

```

**Comment by proddy** [comment]
@MichaelDvP okay to merge and fix things as we find them?

**Comment by MichaelDvP** [comment]
>  Tasmota have their own light HTTPClient library but i couldn't see how they got around not using SSL, so gave up. 

Afair it's a adapted version of BearSSL.
> okay to merge and fix things as we find them?

Have not completly checked the latest changes, but all i've tested works well now, i think it's ok to merge.

**Comment by MichaelDvP** [comment]
grafik

Download link points the the stable version. I think in case of running the latest version it's better to link to the download page https://github.com/emsesp/EMS-ESP32/releases and let the user decide which version he wants.

**Comment by proddy** [comment]
> grafik
> 
> Download link points the the stable version. I think in case of running the latest version it's better to link to the download page https://github.com/emsesp/EMS-ESP32/releases and let the user decide which version he wants.

That was a bug. Hopefully fixed now.

**Pull request merged**